File: app/Shared/ServerDeployer.swift
import Foundation

enum ServerDeployerError: LocalizedError {
    case missingHost
    case missingPassword
    case missingDeployKey
    case sshFailed(String)
    case emptyResponse
    case invalidProfile(String)

    var errorDescription: String? {
        switch self {
        case .missingHost: return "Host is required."
        case .missingPassword: return "Password is required."
        case .missingDeployKey: return "Select a deploy key from the library."
        case .sshFailed(let detail): return "SSH failed: \(detail)"
        case .emptyResponse: return "Server install did not return a profile JSON."
        case .invalidProfile(let detail): return "Invalid profile JSON: \(detail)"
        }
    }
}

enum ServerDeployAuth {
    case password(String)
    case deployKey(DeploySSHKey)
}

enum ServerDeployer {
    struct Result {
        let profile: HopNodeProfile
        let newDeployKey: DeploySSHKey?
    }

    static func deploy(
        host: String,
        port: Int = HopConstants.defaultSSHPort,
        user: String = "root",
        installDir: String = HopConstants.defaultInstallDir,
        auth: ServerDeployAuth,
        onLog: (@Sendable (String) -> Void)? = nil
    ) async throws -> Result {
        let trimmedHost = host.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedHost.isEmpty else { throw ServerDeployerError.missingHost }

        let trimmedUser = user.trimmingCharacters(in: .whitespacesAndNewlines)
        let effectiveUser = trimmedUser.isEmpty ? "root" : trimmedUser

        var newKey: DeploySSHKey?
        let deployPrivateKey: String
        let connectPassword: String?

        switch auth {
        case .password(let password):
            guard !password.isEmpty else { throw ServerDeployerError.missingPassword }
            let generated = try SSHKeyGenerator.generateEd25519(comment: "hopper-deploy@\(trimmedHost)")
            let key = DeploySSHKey(
                name: "Deploy \(effectiveUser)@\(trimmedHost)",
                privateKey: generated.privateKeyPEM
            )
            newKey = key
            deployPrivateKey = generated.privateKeyPEM
            connectPassword = password
        case .deployKey(let key):
            guard !key.privateKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                throw ServerDeployerError.missingDeployKey
            }
            deployPrivateKey = key.privateKey
            connectPassword = nil
        }

        let publicLine: String
        do {
            publicLine = try SSHKeyGenerator.publicKeyLine(
                privateKeyPEM: deployPrivateKey,
                comment: "deploy-\(trimmedHost)"
            )
        } catch {
            throw ServerDeployerError.sshFailed("Could not derive deploy public key.")
        }

        let output: String
        do {
            output = try await DeploySSH.runInstall(
                host: trimmedHost,
                port: port,
                user: effectiveUser,
                password: connectPassword,
                privateKeyPEM: deployPrivateKey,
                publicKeyLine: publicLine,
                installDir: installDir,
                onLog: onLog
            )
        } catch {
            throw ServerDeployerError.sshFailed(error.localizedDescription)
        }

        guard let jsonLine = extractJSONLine(from: output) else {
            throw ServerDeployerError.emptyResponse
        }

        let profile: HopNodeProfile
        do {
            profile = try HopQRParser.parse(jsonLine)
        } catch {
            throw ServerDeployerError.invalidProfile(error.localizedDescription)
        }

        return Result(profile: profile, newDeployKey: newKey)
    }

    private static func extractJSONLine(from output: String) -> String? {
        for line in output.split(separator: "\n", omittingEmptySubsequences: false).reversed() {
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.hasPrefix("{") {
                return trimmed
            }
        }
        return nil
    }
}

File: app-android/app/src/main/java/com/aengix/hopper/provision/ServerDeployer.kt
package com.aengix.hopper.provision

import com.aengix.hopper.data.HopQRParser
import com.aengix.hopper.model.DeploySSHKey
import com.aengix.hopper.model.HopConstants
import com.aengix.hopper.model.HopNodeProfile
import com.aengix.hopper.ssh.DeploySSH
import com.aengix.hopper.ssh.SSHKeyGenerator

sealed class ServerDeployerException(message: String) : Exception(message) {
    data object MissingHost : ServerDeployerException("Host is required.")
    data object MissingPassword : ServerDeployerException("Password is required.")
    data object MissingDeployKey : ServerDeployerException("Select a deploy key from the library.")
    data class SshFailed(val detail: String) : ServerDeployerException("SSH failed: $detail")
    data object EmptyResponse : ServerDeployerException("Server install did not return a profile JSON.")
    data class InvalidProfile(val detail: String) : ServerDeployerException("Invalid profile JSON: $detail")
}

sealed class ServerDeployAuth {
    data class Password(val password: String) : ServerDeployAuth()
    data class DeployKey(val key: DeploySSHKey) : ServerDeployAuth()
}

object ServerDeployer {
    data class Result(
        val profile: HopNodeProfile,
        val newDeployKey: DeploySSHKey?,
    )

    fun deploy(
        host: String,
        port: Int = HopConstants.DEFAULT_SSH_PORT,
        user: String = "root",
        installDir: String = HopConstants.DEFAULT_INSTALL_DIR,
        auth: ServerDeployAuth,
    ): Result {
        val trimmedHost = host.trim()
        if (trimmedHost.isEmpty()) throw ServerDeployerException.MissingHost

        val effectiveUser = user.trim().ifEmpty { "root" }

        val newKey: DeploySSHKey?
        val deployPrivateKey: String
        val password: String?

        when (auth) {
            is ServerDeployAuth.Password -> {
                if (auth.password.isEmpty()) throw ServerDeployerException.MissingPassword
                val generated = SSHKeyGenerator.generateEd25519("hopper-deploy@$trimmedHost")
                newKey = DeploySSHKey(
                    name = "Deploy $effectiveUser@$trimmedHost",
                    privateKey = generated.privateKeyPem,
                )
                deployPrivateKey = generated.privateKeyPem
                password = auth.password
            }
            is ServerDeployAuth.DeployKey -> {
                if (auth.key.privateKey.trim().isEmpty()) throw ServerDeployerException.MissingDeployKey
                newKey = null
                deployPrivateKey = auth.key.privateKey
                password = null
            }
        }

        val publicLine = runCatching {
            SSHKeyGenerator.publicKeyLine(deployPrivateKey, "deploy-$trimmedHost")
        }.getOrElse {
            throw ServerDeployerException.SshFailed("Could not derive deploy public key.")
        }

        val output = runCatching {
            DeploySSH.runInstall(
                host = trimmedHost,
                port = port,
                user = effectiveUser,
                password = password,
                privateKeyPem = deployPrivateKey,
                publicKeyLine = publicLine,
                installDir = installDir,
            )
        }.getOrElse { error ->
            throw ServerDeployerException.SshFailed(error.message ?: error.toString())
        }

        val jsonLine = extractJsonLine(output) ?: throw ServerDeployerException.EmptyResponse

        val profile = runCatching { HopQRParser.parse(jsonLine) }.getOrElse { error ->
            throw ServerDeployerException.InvalidProfile(error.message ?: "parse failed")
        }

        return Result(profile = profile, newDeployKey = newKey)
    }

    private fun extractJsonLine(output: String): String? =
        output.lineSequence().toList().asReversed().firstOrNull { line ->
            line.trim().startsWith("{")
        }?.trim()
}
